estimagic/inference/moment_covs.py

```diff
@@ -0,0 +1,58 @@
+"""
+This module provides functions for the calculation of standard errors for anon linear
+GMM estimator. The source of these calculations is
+Bruce E. Hansen - Econometrics (https://www.ssc.wisc.edu/~bhansen/econometrics).
+"""
+import numpy as np
+
+
+def covariance_moments(mom_cond):
+    """
+    Calculate the standard covariance matrix Omega.
+
+    Args:
+        mom_cond (np.array): 2d array matrix of the moment conditions of
+            dimension (nobs, nmoms).
+
+    Returns:
+        cov (np.array): 2d array covariance matrix of the moments (nmoms, nmoms)
+
+    """
+
+    dev = mom_cond - np.mean(mom_cond, axis=0)
+    cov = dev.T @ dev / mom_cond.shape[0]
+    return cov
+
+
+def gmm_cov(mom_cond, mom_cond_jacob, mom_weight):
+    """
+
+    Args:
+        mom_cond (np.array): 2d array matrix of the moment conditions of
+            dimension (nobs, nmoms).
+
+        mom_cond_jacob (np.array): 3d array of the moment condition derivatives
+            w.r.t. the parameters of dimension (nobs, nmoms, nparams).
+
+        mom_weight (np.array):
+            2d array weighting matrix for the moments of dimension (nmoms, nmoms)
+
+    Returns:
+        sandwich (np.array):
+        2d array variance-covariance matrix of the GMM estimator of dimension
+        (nparams, nparams)
+
+    """
+
+    # Use notation from Hansen book everywhere. Tell flake8 to ignore capital notation
+    W = mom_weight  # noqa: N806
+    Omega = covariance_moments(mom_cond)  # noqa: N806
+    Q = np.mean(mom_cond_jacob, axis=0)  # noqa: N806
+
+    return sandwich_cov(Q, W, Omega, mom_cond.shape[0])
+
+
+def sandwich_cov(Q, W, Omega, nobs):  # noqa: N803
```

Comment:
I would just call this `_make_sandwich`, but if you disagree you can leave it as it is.

Reply:
I sticked to your notation.

Reply:
It definitely has to be public. `gmm_cov` includes additionally the sample dimension, which is not what you want if you are doing msm, where the `nobs`-dimension is already taken care of. 

I can see the benefit for leaving `gmm_cov` as is, although I am not sure whether that 3-d arrays are ever constructed or whether the user would have to set them up. 

In any case, the naming should be improved, but I had no great ideas last week.

Reply:
Ok, then we make it public. I will have to get more practical experience with MSM before I can really contribute to the discussion.

Reply:
Just look at the what `gmm_cov` does before the call to sandwich. Just aggregating over the observations. And of course this is how it is written in a textbook. 

For GMM, the question is whether that aggregation is a feature or an inconvenience from the point of view what is already calculated. I start leaning towards the latter...

For MSM, the moments are pretty much defined without reference to the dimension of the sample, so it clearly is not usable.